doc/architectural/central-data-structures.md

```diff
@@ -0,0 +1,91 @@
+# Central Data Structures
+
+The following is some light technical documentation of the major data structures
+represented in the input transformation to Intermediate Representation (IR) inside
+CBMC and the assorted CProver tools.
+
+## goto_modelt
+
+The `goto_modelt` is the main data structure that CBMC (and the other tools) use to
+represent GOTO-IR (the `GOTO` Intermediate Representation).
+
+A `goto_modelt` is effectively a product type, consisting of:
+
+* A list of GOTO functions (pseudocode: `type goto_functionst = list<goto_functiont>`)
+* A symbol table containing symbol references for the symbols contained in the
+  GOTO functions (pseudocode: `type symbol_tablet = map<identifier, symbolt>`).
+
+The abstract interface of `goto_modelt` is outlined in the file
+[`src/goto-programs/abstract_goto_model.h`](../../src/goto-programs/abstract_goto_model.h).
+
+## goto_functiont
+
+A `goto_functiont` is also a product type. It's designed to represent a function
+at the IR level, and effectively it's the following ADT (in pseudocode):
+
+```js
+type goto_functiont {
```

Comment:
Is there a way to link this to the actual code or generated doxygen?

Reply:
I can link to the file using raw markdown, but I don't think it can link to an actual code structure.

I need to look more into doxygen and its linking capabilities.

Reply:
If you can link it, it might help increase the longevity.  See https://github.com/diffblue/cbmc/pull/7470#issuecomment-1377631947 for my experiences with the lifespan of documentation.  The last thing I want is for you to put in the effort documenting things and then, in 2 years, have to do it all again.

Reply:
In the final documentation, `doxygen` is automatically linking class names to the class documentation it produces, so when the class names are written in full they get automatically picked up by it and linked.

However, given that these are central data structures and calcified by now (i.e. not experiencing significant change with regards to the actual components of the class), I'd rather the pseudocode representation stays as a fast-track way for someone to understand the essence of these data structures without the informational load and visual clutter that the class documentation pages in `doxygen` have.